src/lib.rs

```diff
@@ -13,122 +13,198 @@ use crate::onion_messenger::{run_onion_messenger, MessengerUtilities};
 use bitcoin::secp256k1::PublicKey;
 use home::home_dir;
 use lightning::ln::peer_handler::IgnoringMessageHandler;
-use lightning::onion_message::{DefaultMessageRouter, OnionMessenger};
+use lightning::offers::offer::Offer;
+use lightning::onion_message::{
+    DefaultMessageRouter, OffersMessage, OffersMessageHandler, OnionMessenger, PendingOnionMessage,
+};
 use log::{error, info, LevelFilter};
 use log4rs::append::console::ConsoleAppender;
 use log4rs::append::file::FileAppender;
 use log4rs::config::{Appender, Config as LogConfig, Root};
 use log4rs::encode::pattern::PatternEncoder;
 use std::collections::HashMap;
 use std::str::FromStr;
+use std::sync::Mutex;
 use tonic_lnd::lnrpc::GetInfoRequest;
+use triggered::{Listener, Trigger};
 
 pub struct Cfg {
     pub lnd: LndCfg,
     pub log_dir: Option<String>,
+    // Use to externally trigger shutdown.
+    pub shutdown: Trigger,
+    // Used to listen for the signal to shutdown.
+    pub listener: Listener,
 }
 
-pub async fn run(args: Cfg) -> Result<(), ()> {
-    let log_dir = args.log_dir.unwrap_or_else(|| {
-        home_dir()
-            .unwrap()
-            .join(".lndk")
-            .join("lndk.log")
-            .as_path()
-            .to_str()
-            .unwrap()
-            .to_string()
-    });
-
-    // Log both to stdout and a log file.
-    let stdout = ConsoleAppender::builder().build();
-    let lndk_logs = FileAppender::builder()
-        .encoder(Box::new(PatternEncoder::new("{d} - {m}{n}")))
-        .build(log_dir)
-        .unwrap();
-
-    let config = LogConfig::builder()
-        .appender(Appender::builder().build("stdout", Box::new(stdout)))
-        .appender(Appender::builder().build("lndk_logs", Box::new(lndk_logs)))
-        .build(
-            Root::builder()
-                .appender("stdout")
-                .appender("lndk_logs")
-                .build(LevelFilter::Info),
-        )
-        .unwrap();
-
-    let _log_handle = log4rs::init_config(config).unwrap();
+pub enum OfferError {
+    OfferAlreadyAdded,
+}
 
-    let mut client = get_lnd_client(args.lnd).expect("failed to connect");
+enum OfferState {
+    OfferAdded,
+    InvoiceRequestSent,
+    InvoiceReceived,
+    InvoicePaymentDispatched,
+    InvoicePaid,
+}
 
-    let info = client
-        .lightning()
-        .get_info(GetInfoRequest {})
-        .await
-        .expect("failed to get info")
-        .into_inner();
+pub struct OfferHandler {
```

Comment:
I'm trying to think through a name that'll encompass that this struct is used both for forwarding onion messages (which might not be related to offers) and handling offer specific message. What about naming this something like... InternalOnionMessenger or OnionHandler so it doesn't conflict with LDK's OnionMessenger? 

No rush to reply to this, I know you're on vacation haha. Just want to keep track of some of the questions I have here in the meantime...